### PR TITLE
Dk 4

### PR DESCRIPTION
**Pull request by danielakarwowska**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+const express = require("express");
```
Comment (pdybowski) [inline app.js:1]: we dont use such a file

File: config/default.json [hunk 2]
```diff
@@ -1,4 +1,4 @@
 {
-    "jwtPrivateKey": "",
+    "jwtPrivateKey": "321",
```
Comment (pdybowski) [inline config/default.json:2]: you can remove whole config folder since we use .env file

File: controllers/AuthConstrollers.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        return res.status(409).json({ message: "User with email already exists!" });
+      }
+    
+      const newUser = new User({ username, email, password,phoneNumber:req.body.phoneNumber });
+      const savedUser = await newUser.save().catch((err) => {
+        console.log("Error: ", err);
+        res.status(500).json({ error: "Cannot register user at the moment!" });
+      });
+    
+      if (savedUser) res.set("x-auth-token","registerToken").json({ message: "Thanks for registering" });
```
Comment (pdybowski) [inline controllers/AuthConstrollers.js:31]: everything execept successfull res.send(...) should be placed in service

File: controllers/AuthConstrollers.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if (userWithEmail.password !== password)
+    return res
+      .status(400)
+      .json({ message: "Email or password does not match!" });
+      const jwtToken = jwt.sign(
+        { id: userWithEmail.id, email: userWithEmail.email },
+        process.env.JWT_SECRET
+      );
+    
+          res.json({ message: "Welcome Back!", token: jwtToken });
```
Comment (pdybowski) [inline controllers/AuthConstrollers.js:60]: same

File: controllers/AuthConstrollers.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        findOneAndDelete({User:req.User, type:"login"}, (err,doc)=>{
+          if(err)return res.status(401).json({
+            status:false,
+            message:"Server error, logout failed",
+          })
+          return res.status(200).json({
+            status:true,
+            message:"Logout succesfully"
+          })
+        })
```
Comment (pdybowski) [inline controllers/AuthConstrollers.js:73]: dont understand .we should just send request to db. mayyyybe check if user exists and send 200 back to front

File: controllers/AuthConstrollers.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
+const router = require("express").Router();
```
Comment (pdybowski) [inline controllers/AuthConstrollers.js:1]: why do we have to controller files?

File: controllers/auth.controller.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 const authService = require("../services/auth.service");
+const AuthControlers = require ("../controllers/AuthConstrollers")
```
Comment (pdybowski) [inline controllers/auth.controller.js:2]: AuthController (without 's'). its just one controller that controls auth rotes

File: index.js [hunk 8]
```diff
@@ -1,10 +1,20 @@
...
 
-
+app.get("/", (req,res) =>{
+    res.json({
+        message:"12345",
+    })
+})
+// app.use ('auth', require ("./routers/auth"))
+// app.use ("/user", UserRouters);
+// app.use("/api", AuthRoute)
```
Comment (pdybowski) [inline index.js:17]: remove lol

File: models/user.model.js [hunk 9]
```diff
@@ -8,46 +8,46 @@ const userSchema = Schema({
     trim: true,
     minlength: 2,
     maxlength: 20,
-    required: true,
+    // required: true,
```
Comment (pdybowski) [inline models/user.model.js:11]: ?

File: app.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+mongoose
+  .connect(process.env.MONGO_URL)
+  .then(console.log("Connected to MONGODB"))
+  .catch((err) => console.log(err));
+
+app.get("/", function (req, res) {
+  res.send("Hello!");
+});
+console.log('DANIELA');
+app.use("/routers", authRoute);
```
Comment (pdybowski) [inline app.js:17]: app.user("auth", ...) but.. still not in this file

File: routers/auth.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+const router = require("express").Router();
+const User = require("../models/user.model")
+const bcrypt = require ("bcryptjs")
+const jwt = require("jsonwebtoken")
+require ("dotenv").config();
```
Comment (pdybowski) [inline routers/auth.js:5]: do we need it here?

File: routers/auth.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+const router = require("express").Router();
+const User = require("../models/user.model")
+const bcrypt = require ("bcryptjs")
+const jwt = require("jsonwebtoken")
```
Comment (pdybowski) [inline routers/auth.js:4]: do we need it here?

File: routers/auth.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+const router = require("express").Router();
+const User = require("../models/user.model")
+const bcrypt = require ("bcryptjs")
```
Comment (pdybowski) [inline routers/auth.js:3]: do we need it here?

File: routers/auth.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+const router = require("express").Router();
+const User = require("../models/user.model")
```
Comment (pdybowski) [inline routers/auth.js:2]: do we need it here?

File: routers/auth.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+const bcrypt = require ("bcryptjs")
+const jwt = require("jsonwebtoken")
+require ("dotenv").config();
+
+const AuthControlers = require ("../controllers/AuthConstrollers")
+
+
+
+router.post('/Register', AuthControlers.register)
```
Comment (pdybowski) [inline routers/auth.js:11]: /register

File: routers/auth.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+
+router.post('/Login',AuthControlers.login)
```
Comment (pdybowski) [inline routers/auth.js:14]: /login

File: routers/auth.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+
+
+router.get('/logout', AuthControlers.logout)
```
Comment (pdybowski) [inline routers/auth.js:17]: logout

File: services/auth.service.js [hunk 18]
```diff
@@ -5,7 +5,6 @@ const {
   InernalServerError,
 } = require("../helpers/errorHandlers");
 const { hashInput, hashCompare } = require("../hash");
-
 let { refreshTokens } = undefined; //TODO should be from database
```
Comment (pdybowski) [inline services/auth.service.js:8]: remove that line. we dont use that token for now

File: views/Login.ejs [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+    <title>Document</title>
+</head>
+<body>
+    <h1>LOGIN</h1>
+    <form id="-form">
+        <input type="text" id="username" placeholder="Username"/>
+        <input type="password" id="password" placeholder="password"/>
+        <input type="submit" value="Submit Form"/> 
+        </form>
+</body>
```
Comment (pdybowski) [inline views/Login.ejs:16]: remova that file

File: views/SignUp.ejs [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+    <title>Document</title>
+</head>
+<body>
+    <h1>SignUp</h1>
+    <form id="SignUp-form">
+        <input type="text" id="username" placeholder="Username"/>
+        <input type="password" id="password" placeholder="password"/>
+        <input type="submit" value="Submit Form"/> 
+        </form>
+</body>
```
Comment (pdybowski) [inline views/SignUp.ejs:16]: remove that file

**Review by pdybowski — CHANGES_REQUESTED** [review]
please fix

File: controllers/auth.controller.js [hunk 21]
```diff
@@ -1,17 +1,43 @@
...
 
-exports.login = (req, res, next) => {
-  try {
-    const data = await authService.login(req.body); // TODO probably broken - its just an example
-    return responseWithToken(res, data); //TODO
-  } catch (error) {
+exports.register = async (req, res, next) =>{
+  try{
+    const token = AuthService.register(req.body);
+    res.set("x-auth-token", token).json({message: "Tranks for registering"});
```
Comment (pdybowski) [inline controllers/auth.controller.js:13]: res.header("X-Auth-Token", data.token).json(data);

where data is an object containing user data:
token, 
isAdmin (from user model)
name (from user model)

File: controllers/auth.controller.js [hunk 22]
```diff
@@ -1,17 +1,43 @@
...
+  }
+};
+
+exports.logout = async (req,res,next) =>{
+  try{
+    const token = AuthService.logout(req.body);
+    res.set("x-auth-token", token).json({
+      status: true,
+      message: "Logout succesfully",
+    });
```
[the remark this answers is not in the document]
Comment (pdybowski) [reply controllers/auth.controller.js:34]: res.status(200).send(true)

File: services/auth.service.js [hunk 23]
```diff
@@ -17,27 +15,42 @@ generateAccessToken = async (user) => {
...
+    throw new BadRequestError ("User already registered");
+  }
+  const newUser = new User({
+    username,
+    email,
+    password,
+    phoneNumber:phoneNumber,
+  });
+  await newUser.save();
+  return generateAccessToken;
```
Comment (pdybowski) [inline services/auth.service.js:34]: const resUser = _.pick(newUser, [
"_id",
"isAdmin",
"name"
  ]);

return { token,  reqUser }

File: services/auth.service.js [hunk 24]
```diff
@@ -17,27 +15,42 @@ generateAccessToken = async (user) => {
...
+  return generateAccessToken
 };
+
+//logout
+exports.logout = async(body) =>{
+  const{email,password} = body;
+  const user = await User.findOne({ email: email });
+  if(!user){
+    throw new BadRequestError
+    ("User doesn't exist")
```
Comment (pdybowski) [inline services/auth.service.js:54]: put message in the same line as error

File: services/auth.service.js [hunk 25]
```diff
@@ -17,27 +15,42 @@ generateAccessToken = async (user) => {
...
+//login
+exports.login = async(body) =>{
+  const{email} = body;
+  const userWithEmail = await User.findOne({ email: email });
 
-  return { accessToken: accessToken, refreshToken: refreshToken }; //TODO probably we we should return also userRole
+  if (!userWithEmail){
+    throw new BadRequestError ("User doesn't exist!")
+  }
+  return generateAccessToken
```
Comment (pdybowski) [inline services/auth.service.js:45]: return { token, isAdmin, name, _id }

File: controllers/auth.controller.js [hunk 26]
```diff
@@ -1,17 +1,43 @@
...
     next(error);
   }
 };
 
-// TODO register
-// TODO logout
+exports.login = async (req, res, next) =>{
+  try{
+    const token = AuthService.login(req.body);
+    res.set("x-auth-token", token).json({message:"Welcome back"});
```
Comment (pdybowski) [inline controllers/auth.controller.js:22]: same as above

**Review by pdybowski — CHANGES_REQUESTED** [review]
please fix

File: controllers/auth.controller.js (hunk 21 above)
Comment (danielakarwowska) [reply controllers/auth.controller.js:13]: _id

**Review by pdybowski — APPROVED** [review]

**Pull request merged**